### PR TITLE
improve: Also pass along a state (nonce) paramater.

### LINKED ISSUE #1134: OIDC login fails due to Authentik sending state QS and node-openid-client failing a check
It's all here.  https://forum.getodk.org/t/oidc-issue-when-logging-in/46619

### PR DESCRIPTION
**Pull request by openbrian**
bug: #1134

It's not needed for CSRF protection because code_challenge does this, but the Authentik IDP will return &state= in the query string.  This will trigger node-openid-client to fail a check.

Closes #1134 

#### Before submitting this PR, please make sure you have:

- [ ] run `make test` and confirmed all checks still pass OR confirm CircleCI build passes
- [ ] verified that any code from external sources are properly credited in comments or that everything is internally sourced

Testing the oidc way failed for me.  I mean the tests would not kick off.

### CODE REVIEW

File: lib/resources/oidc.js [hunk 1]
```diff
@@ -102,15 +103,19 @@ module.exports = (service, endpoint) => {
       const client = await getClient();
       const code_verifier = generators.codeVerifier(); // eslint-disable-line camelcase
 
+      // State is not strictly required because PKCS (code_challenge) protects us from CSRF attacks.
```
Comment (lognaturel) [inline lib/resources/oidc.js:106]: It feels important to then mention why it's there! Maybe just add "but some IDPs always send it"

**Comment by lognaturel** [comment]
Thanks for digging into this and figuring out a solution, @openbrian! I've left one small suggestion. @alxndrsn could you please do a deeper review?

File: lib/resources/oidc.js (hunk 1 above)
Comment (openbrian) [reply lib/resources/oidc.js:106]: done

File: lib/resources/oidc.js [hunk 2]
```diff
@@ -102,15 +103,22 @@ module.exports = (service, endpoint) => {
...
+      // See https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#name-countermeasures-4
+      // We create a nonce (state) anyway because some IDPs (e.g. Authentik) will send the nonce
```
Comment (alxndrsn) [inline lib/resources/oidc.js:108]: `nonce` is the name of a different value which can be used at various points in OAuth/OIDC.  I think avoiding that word here would reduce chance of confusion.

File: lib/resources/oidc.js [hunk 3]
```diff
@@ -102,15 +103,22 @@ module.exports = (service, endpoint) => {
       const client = await getClient();
       const code_verifier = generators.codeVerifier(); // eslint-disable-line camelcase
 
+      // State is not strictly required because PKCS (code_challenge) protects us from CSRF attacks.
```
Comment (alxndrsn) [inline lib/resources/oidc.js:106]: Should `PKCS` be `PKCE`?

**Comment by alxndrsn** [comment]
> It's not needed for CSRF protection because code_challenge does this, but the Authentik IDP will return &state= in the query string. This will trigger node-openid-client to fail a check.

Thanks for the PR!  I think this is a bug in Authentik's implementation; they seem to be open source (https://github.com/goauthentik/authentik?) so perhaps it can be fixed on their side.

That said, `state` _should_ be supported by all IdPs (per the spec).  Also `state` reportedly adds some security against generating fake errors.  I will dig deeper and test this against some different IdPs.

File: lib/resources/oidc.js (hunk 3 above)
Comment (openbrian) [reply lib/resources/oidc.js:106]: Yes, I'm on it ;)

File: lib/resources/oidc.js [hunk 4]
```diff
@@ -102,15 +103,22 @@ module.exports = (service, endpoint) => {
...
+      // See https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#name-countermeasures-4
+      // We create a nonce (state) anyway because some IDPs (e.g. Authentik) will send the nonce
+      // back to us.  If it doesn't explicitly get one, it sends back an empty string.
+      const state = generators.state(16);
```
[the remark this answers is not in the document]
Comment (openbrian) [reply lib/resources/oidc.js:110]: I played in the playground here https://www.oauth.com/playground/authorization-code-with-pkce.html and they had a 16 byte state.  There's like zero cost to use 32 bytes.  It's not encoding or signing anything.  I'll raise it to 32, aka use the default.

File: lib/resources/oidc.js (hunk 2 above)
Comment (alxndrsn) [reply lib/resources/oidc.js:108]: @openbrian just in case you missed this :slightly_smiling_face:

**Comment by alxndrsn** [comment]
Related: https://github.com/panva/openid-client/discussions/703

**Pull request closed without merging**